### PR TITLE
Add server tracing with RestEasyExample

### LINKED ISSUE #2: Server side JAX-RS tracing
* Provide an auto-instrumentation for server-side JAX-RS
* The instrumentation should use OpenTelemetry API artifact

SmallRyeOpenTracing is using JAX-RS request filter to start spans and a servlet filter to finish spans. The servlet filter is necessary to catch exceptions and finish spans exceptions.

OpenTracing implementation lives here https://github.com/opentracing-contrib/java-jaxrs/tree/master/opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server

### PR DESCRIPTION
**Pull request by pavolloffay**
Signed-off-by: Pavol Loffay <ploffay@redhat.com>

Resolves #2 

This PR adds JAX-RS server-side tracing implementation. The TCK will be added in https://github.com/smallrye/smallrye-opentelemetry/issues/19. 

This implementation is intended to be used in runtimes supporting JAX-RS e.g. Wildfly. The runtimes that provide JAX-RS compatible API but the underlying implementation is not servlet based could use a different implementation to start and finish spans, however JAX-RS bindings might required to set correct operation names - FQCN, or values frorm `@Path` annotations. In this case the JAX-RS layer could just annotate spans with the required information.

### CODE REVIEW

**Comment by kenfinnigan** [comment]
I'm still not convinced we really need a `smallrye-opentelemetry`, or an MP spec it implements.

We also need to split things out into more modules as running on Quarkus would not use JAX-RS filters, but the underlying Vert.x ones.

Will try and look at this more over the coming week when I'm looking at adding OpenTelemetry to Quarkus

**Comment by pavolloffay** [comment]
> I'm still not convinced we really need a smallrye-opentelemetry, or an MP spec it implements.

How do you want to provide tracing capabilities in the MicroProfile then? OpenTelemetry could most likely provide a unified way to collector other telemetry signals not just traces. 

 > We also need to split things out into more modules as running on Quarkus would not use JAX-RS filters, but the underlying Vert.x ones.

The PR implements JAX-RS server filters, another implementation can be added in a separate module. What we need more is a TCK that can be used across different runtimes. That will come in a separate PR.

**Comment by kenfinnigan** [comment]
OpenTelemetry can provide tracing capabilities, without there being MP APIs or custom implementations.

It might just be as simple as a specification that says do X with OpenTelemetry to achieve Y

**Comment by pavolloffay** [comment]
> OpenTelemetry can provide tracing capabilities, without there being MP APIs or custom implementations.

We don't want a custom implementation of the API or even worse the SDK. What you have mentioned is what MP-OpenTracing was doing (e.g. automatic instrumentation) and that we want to provide in the MP-OpenTelemetry project.

**Comment by pavolloffay** [comment]
@kenfinnigan could we merge this and start working on the TCK? If the Quarkus will use instrumentation at lower layers we could host it in this repository in a separate maven artifact. This PR adds `smallrye-opentelemetry-jaxrs2` that use JAX-RS filters.

File: examples/resteasy/src/main/java/io/smallrye/opentelemetry/example/resteasy/ExampleApplication.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * @author Pavol Loffay
+ */
+public class ExampleApplication extends Application {
+    private final HashSet<Object> singletons = new HashSet<>();
+
+    public ExampleApplication() {
+        TracerSdkProvider tracerProvider = OpenTelemetrySdk.getTracerProvider();
+        TracerSdkProvider.builder().build();
```
Comment (kenfinnigan) [inline examples/resteasy/src/main/java/io/smallrye/opentelemetry/example/resteasy/ExampleApplication.java:24]: Does the return value not need to be used in some way?

Seems odd to build something that doesn't appear to be stored anywhere inside those calls and then discard it

File: implementation-jaxrs2/src/main/java/io/smallrye/opentelemetry/jaxrs2/server/ServerTracingInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import io.opentelemetry.trace.Tracer;
+import io.smallrye.opentelemetry.jaxrs2.serialization.InterceptorSpanDecorator;
+import io.smallrye.opentelemetry.jaxrs2.serialization.TracingInterceptor;
+import io.smallrye.opentelemetry.jaxrs2.server.internal.CastUtils;
+import io.smallrye.opentelemetry.jaxrs2.server.internal.SpanWrapper;
+
+@Priority(Priorities.ENTITY_CODER)
+public class ServerTracingInterceptor extends TracingInterceptor {
+    /**
```
Comment (kenfinnigan) [inline implementation-jaxrs2/src/main/java/io/smallrye/opentelemetry/jaxrs2/server/ServerTracingInterceptor.java:21]: What implementation is there that's using CXF?

File: examples/resteasy/src/main/java/io/smallrye/opentelemetry/example/resteasy/ExampleApplication.java (hunk 1 above)
Comment (pavolloffay) [reply examples/resteasy/src/main/java/io/smallrye/opentelemetry/example/resteasy/ExampleApplication.java:24]: This seems like a leftover. I will remove it.

File: implementation-jaxrs2/src/main/java/io/smallrye/opentelemetry/jaxrs2/server/ServerTracingInterceptor.java (hunk 2 above)
Comment (pavolloffay) [reply implementation-jaxrs2/src/main/java/io/smallrye/opentelemetry/jaxrs2/server/ServerTracingInterceptor.java:21]: This instrumenation is based on https://github.com/opentracing-contrib/java-jaxrs that was compatible and tested against RestEasy, Jersey and Apache CXF. 

Once this POC is completed we could consider moving the JAX-RS instrumentation to OpenTelemetry project and just reuse it here like we did in SmallRye-OpenTracing.

Comment (pavolloffay) [reply implementation-jaxrs2/src/main/java/io/smallrye/opentelemetry/jaxrs2/server/ServerTracingInterceptor.java:21]: Also if I am not mistaken OpenLiberty uses CXF and they probably would like to use this implementation for tracing.

File: api/pom.xml [hunk 3]
```diff
@@ -9,22 +9,13 @@
...
-            <groupId>org.jacoco</groupId>
-            <artifactId>jacoco-maven-plugin</artifactId>
-          </plugin>
-        </plugins>
-      </build>
-    </profile>
-  </profiles>
+  <dependencies>
+    <dependency>
+      <groupId>javax.enterprise</groupId>
```
Comment (kenfinnigan) [inline api/pom.xml:16]: This should be the jakarta artifact

Comment (pavolloffay) [reply api/pom.xml]: done updated all artifacts to `jakarta`.

File: implementation-jaxrs2/src/main/java/io/smallrye/opentelemetry/jaxrs2/server/ServerTracingInterceptor.java (hunk 2 above)
Comment (kenfinnigan) [reply implementation-jaxrs2/src/main/java/io/smallrye/opentelemetry/jaxrs2/server/ServerTracingInterceptor.java:21]: "probably would like" is not actual usage.

Can you confirm with them whether they would use this. If they won't, I'd prefer we remove any CXF handling

**Pull request merged**